### PR TITLE
Resolve #6 Parse/Stringify Auction

### LINKED ISSUE #6: Parsing/stringifying an auction
We want some way of being able to convert an auction to and from a string. The simplest format is probably something like `1S--2S-X-XX---3H----` where calls are separated by the `-` character and pass is optional.

### PR DESCRIPTION
**Pull request by KripaPanchagnula**
There are some edge cases about uncontested auctions that we may want to improve stylistically before merging

### CODE REVIEW

File: src/string-parser/parse-auction.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { parseCall } from '.';
+import { AuctionCall } from '../types';
+/**
+ * Converts a contested auction string into a list of AuctionCalls
+ * @param auction String of calls in auction separated by -
+ * @returns List of AuctionCalls representing the auction
+ */
+function parseContestedAuction(auction: string): AuctionCall[] {
+	const calls = auction.split('-');
+	const bids = [];
```
Comment (aaron-hutton) [inline src/string-parser/parse-auction.ts]: Can we switch this to a `.map`

File: src/string-parser/parse-call.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Parses a single call in an auction. It can either be a bid
+ * @param bid A string of the call to be parsed.
+ * @returns AuctionCall.call object or an error if it can't be parsed
+ */
+
+export function parseCall(bid: string): AuctionCall {
+	if (!Number.isNaN(Number(bid[0]))) {
+		return { call: { level: Number(bid[0]), suit: parseStrain(bid[1]) } };
```
Comment (aaron-hutton) [inline src/string-parser/parse-call.ts]: Need to check the level is 1-7

File: src/string-parser/parse-call.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * Checks strain of bid for NT or a suit
+ * @param str Letter denoting suit or NT
+ * @returns NT or Suit type
+ */
+function parseStrain(str: string): Types.Suit | NoTrumpType {
+	return str === 'n' || str === 'N' ? Types.NoTrump : parseSuit(str);
+}
+
+/**
+ * Parses a single call in an auction. It can either be a bid
```
Comment (aaron-hutton) [inline src/string-parser/parse-call.ts]: You didn't finish this comment

File: src/string-parser/parse-call.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * @param bid A string of the call to be parsed.
+ * @returns AuctionCall.call object or an error if it can't be parsed
+ */
+
+export function parseCall(bid: string): AuctionCall {
+	if (!Number.isNaN(Number(bid[0]))) {
```
Comment (aaron-hutton) [inline src/string-parser/parse-call.ts]: This doesn't really check a whole lot, [see here](https://thisthat.dev/number-constructor-vs-parse-int/). It's probably better to just switch to `parseInt`. I'd also only do the parsing once and save it to a variable.

Comment (aaron-hutton) [reply src/string-parser/parse-call.ts]: I think I'd also move the switch statement first then do this `if` check because I think it will make it easier to read

File: src/string-parser/stringify-auction.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import { AuctionCall } from '../types';
+/**
+ * Converts a list of calls in a contested auction to a string separated by -
+ * @param auction List of calls in auction
+ * @returns String representation of auction
+ */
+function stringifyContestedAuction(auction: AuctionCall[]): string {
+	const stringAuction = [];
+	for (const call of auction) {
+		stringAuction.push(stringifyCall(call));
```
Comment (aaron-hutton) [inline src/string-parser/stringify-auction.ts]: I'd switch this to a `.map` as well

File: test/string-parser/parse-auction.test.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	{ call: PossibleCalls.Pass },
+	{ call: { level: 2, suit: Types.Suit.Club } },
+	{ call: PossibleCalls.Pass },
+	{ call: { level: 2, suit: Types.Suit.Heart } },
+	{ call: PossibleCalls.Pass },
+	{ call: { level: 2, suit: Types.NoTrump } },
+	{ call: PossibleCalls.Pass },
+	{ call: { level: 3, suit: Types.Suit.Club } },
+	{ call: PossibleCalls.Pass },
+	{ call: { level: 3, suit: Types.NoTrump } },
```
Comment (aaron-hutton) [inline test/string-parser/parse-auction.test.ts:42]: Hmm, this poses the question of an uncontested auction ending with three passes. e.g. `1H-2H-P` you'd probably expect to have three passes at the end. Maybe the solution is to always add a final pass to the end of uncontested auctions. Come to think of it you may want some way of starting with 1 initial pass in an uncontested auction, which isn't currently possible.

File: test/string-parser/parse-auction.test.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+];
+
+describe('Testing parseAuction', () => {
```
Comment (aaron-hutton) [inline test/string-parser/parse-auction.test.ts]: `StringParser.parseAuction`

File: src/string-parser/parse-call.ts (hunk 2 above)
Comment (aaron-hutton) [reply src/string-parser/parse-call.ts]: Also can we pass `bid.slice(1)` to `parseStrain` and make that handle `n, N, nt, NT` because otherwise `1NQ` or `2H5` are valid calls which is weird.

**Review by aaron-hutton — CHANGES_REQUESTED** [review]

File: test/string-parser/parse-auction.test.ts (hunk 6 above)
Comment (KripaPanchagnula) [reply test/string-parser/parse-auction.test.ts:42]: Yeah this was one of the edge cases I was worried about. I dont think we need to force an auction to have 3 passes at the end here, because we may want partial auctions e.g. when posing problems?

Comment (KripaPanchagnula) [reply test/string-parser/parse-auction.test.ts:42]: TODO add comment that it returns an extra pass at the end

File: src/string-parser/stringify-auction.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	}
+	return stringAuction.join('-');
+}
+/**
+ * Converts a list of calls in an uncontested auction to a string separated by - with all passes inserted
+ * @param auction List of calls in auction
+ * @returns String representation of auction
+ */
+function stringifyUncontestedAuction(auction: AuctionCall[]): string {
+	const stringUnconstedAuction = stringifyContestedAuction(auction);
```
Comment (KripaPanchagnula) [inline src/string-parser/stringify-auction.ts]: TODO check if there is a final pass and strip it. It needs to take as an argument the result of `parseUncontestedAuction` and return its initial string. Check its an even length and all odd indices are passes.

**Review by aaron-hutton — APPROVED** [review]
